### ISSUE TITLE
How to insert snippets by only typing partial prefix of the trigger word?

### ISSUE THREAD

**Issue by danijar** [SirVer/ultisnips#1207]
I would like to write snippets that can be expanded after typing only a part of the trigger word. For example, `i`, `im`, `imp`, `impo`, `impor` should all expand to `import ` when followed by the trigger key, e.g. <kbd>Tab</kbd>. This is helpful because one may think of a snippet after having already typed a couple of characters of a code construct. At the moment, this is possible through somewhat complicated regexes, making it hard to define such snippets. Is this supported in an easier way? If not, would it make sense to add a snippet option for this?

**Comment by danijar** [comment]
Nevermind, I found #268. You're mentioning completers that are better suited for this use case. Which ones were you thinking of?

**Comment by tbrazel** [comment]
As you noted, the regex for this would be rough: `i(?:m(?:p(?:o(?:r(?:t?)?)?))?)?` Personally I would also love the feature that you're proposing and that was proposed in #268.

One popular completer is [deoplete](https://github.com/Shougo/deoplete.nvim), and many people swear by it.

**Comment by danijar** [comment]
Thanks for sharing the regex. If this worked it should be really easy to add this feature as a snippet option to UltiSnips. I'm using droplete but so far it seems to fill quite a different use case than triggering snippets that may include anchors for the cursor to jump to etc.

**Comment by SirVer** [comment]
Yes, this would not be a hard feature to add. I came around to it as maybe being a good idea to implement and would love to see a patch for it.

**Comment by SirVer** [comment]
Shipped in #1670 as a `p` snippet option. With it, any non-empty prefix of the trigger expands the snippet, and only the characters the user actually typed are consumed:

```snippets
snippet import "imp" p
import $0
endsnippet
```

Now any of `i`, `im`, `imp`, `impo`, `impor` or `import` followed by `<Tab>` expands. Text typed before the prefix is left alone (`hello imp<Tab>` → `hello import `). Plays nicely with the existing flags; combine with `b` to restrict to start of line, etc. See `:h UltiSnips-snippet-options` (option `p`).

**Comment by wookayin** [comment]
That's a great feature, thanks @SirVer. One issue I find is that the `p` option makes every single prefix of the pattern as a trigger:

> If several snippets resolve to the same prefix you will be prompted to choose between them.

which makes sense, but in some cases this might be inconvenient. For example:

```
# a little hypothetical, for demo purposes
snippet i "i?"
i love ultisnips
endsnippet
snippet import "imp" p
import $0
endsnippet
snippet ipython "ipython" p
__import__('IPython').embed()
endsnippet
```

If there were a snippet `i` that is a (proper) prefix of another snippet with the 'p' option, users would probably mean this should be automatically chosen with a higher priority and expand without a confirmation:

```
1: (i) "i?" 
2: (import) "imp" 
3: (ipython) "ipython" 
```

I'm also happy to discuss in a new issue thread if it sounds better. Thanks for your support!

**Comment by SirVer** [comment]
@wookayin I was considering that and my initial thought was similar to yours; however it felt inconsistent with the other rules of UltiSnips - it would hide snippets that would be shown in ListSnippets and in fact should be listed. Your example up there is consistent with all the other rules of ultisnips - if multiple matches, let the user choose.

Also, if indeed a user wants to model what you describe, there is a way by bumping up the priority of this snippet. Example:

```snippets
priority 1
snippet i "i"
i love ultisnips
endsnippet

priority 0
snippet import "imp" p
import $0
endsnippet

snippet ipython "ipython" p
__import__('IPython').embed()
endsnippet
```

Typing `i<Tab>` expands the `i` snippet directly (no prompt), while `im<Tab>` still expands `import` and `ip<Tab>` still expands `ipython`. So the choice is in the snippet author's hands without changing the default rule.
